File: mobile/iosApp/iosApp/SimulationViewModel.swift
import Foundation
import shared

@MainActor
class SimulationViewModel: ObservableObject {
    @Published var isRunning = false
    @Published var selectedScenarioId: String = "rome"
    @Published var singleResult: SimulationResult? = nil
    @Published var analysis: MonteCarloRunner.MonteCarloAnalysis? = nil
    @Published var durationMs: Int64 = 0
    @Published var errorMessage: String? = nil

    var availableScenarios: [ScenarioRegistry.ScenarioInfo] {
        ScenarioRegistry.shared.listAll() as! [ScenarioRegistry.ScenarioInfo]
    }

    var selectedScenarioInfo: ScenarioRegistry.ScenarioInfo? {
        availableScenarios.first { $0.id == selectedScenarioId }
    }

    func selectScenario(_ id: String) {
        selectedScenarioId = id
        singleResult = nil
        analysis = nil
    }

    func runSingle() {
        isRunning = true
        errorMessage = nil
        Task {
            do {
                let scenario = ScenarioRegistry.shared.get(id: selectedScenarioId)
                let start = DispatchTime.now()
                let engine = SimulationEngine(scenario: scenario, baseSeed: scenario.simulationRules.baseRandomSeed)
                let result = engine.run(runIndex: 0)
                let end = DispatchTime.now()
                let nanos = end.uptimeNanoseconds - start.uptimeNanoseconds
                self.durationMs = Int64(nanos / 1_000_000)
                self.singleResult = result
                self.isRunning = false
            } catch {
                self.errorMessage = error.localizedDescription
                self.isRunning = false
            }
        }
    }

    func runMonteCarlo() {
        isRunning = true
        errorMessage = nil
        Task {
            do {
                let scenario = ScenarioRegistry.shared.get(id: selectedScenarioId)
                let start = DispatchTime.now()
                let runner = MonteCarloRunner(scenario: scenario)
                let results = try await runner.runAll()
                let end = DispatchTime.now()
                let nanos = end.uptimeNanoseconds - start.uptimeNanoseconds
                self.durationMs = Int64(nanos / 1_000_000)
                self.analysis = MonteCarloRunner.companion.analyze(results: results)
                self.isRunning = false
            } catch {
                self.errorMessage = error.localizedDescription
                self.isRunning = false
            }
        }
    }
}

File: mobile/androidApp/src/main/java/org/flossware/civilization/android/SimulationViewModel.kt
package org.flossware.civilization.android

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.flossware.civilization.engine.MonteCarloRunner
import org.flossware.civilization.engine.SimulationEngine
import org.flossware.civilization.engine.SimulationResult
import org.flossware.civilization.scenarios.ScenarioRegistry

data class SimulationUiState(
    val isRunning: Boolean = false,
    val mode: String = "single",
    val selectedScenarioId: String = "rome",
    val availableScenarios: List<ScenarioRegistry.ScenarioInfo> = ScenarioRegistry.listAll(),
    val singleResult: SimulationResult? = null,
    val monteCarloResults: List<SimulationResult>? = null,
    val analysis: MonteCarloRunner.MonteCarloAnalysis? = null,
    val durationMs: Long = 0,
    val error: String? = null
)

class SimulationViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(SimulationUiState())
    val uiState: StateFlow<SimulationUiState> = _uiState

    fun selectScenario(scenarioId: String) {
        _uiState.value = _uiState.value.copy(
            selectedScenarioId = scenarioId,
            singleResult = null,
            monteCarloResults = null,
            analysis = null
        )
    }

    fun runSingle() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isRunning = true, mode = "single", error = null)
            try {
                val scenario = ScenarioRegistry.get(_uiState.value.selectedScenarioId)
                val startTime = System.currentTimeMillis()
                val engine = SimulationEngine(scenario, scenario.simulationRules.baseRandomSeed)
                val result = engine.run(0)
                val duration = System.currentTimeMillis() - startTime

                _uiState.value = _uiState.value.copy(
                    isRunning = false,
                    singleResult = result,
                    durationMs = duration
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isRunning = false, error = e.message)
            }
        }
    }

    fun runMonteCarlo() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isRunning = true, mode = "monte", error = null)
            try {
                val scenario = ScenarioRegistry.get(_uiState.value.selectedScenarioId)
                val startTime = System.currentTimeMillis()
                val runner = MonteCarloRunner(scenario)
                val results = runner.runAll()
                val duration = System.currentTimeMillis() - startTime
                val analysis = MonteCarloRunner.analyze(results)

                _uiState.value = _uiState.value.copy(
                    isRunning = false,
                    monteCarloResults = results,
                    analysis = analysis,
                    durationMs = duration
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isRunning = false, error = e.message)
            }
        }
    }
}
